include/albatross/src/models/patchwork_gp.hpp

```diff
@@ -0,0 +1,555 @@
+                             const std::vector<GroupKey> &groups) {
+  using BoundarySubFeatureType =
+      typename invoke_result<BoundaryFunction, GroupKey,
+                             GroupKey>::type::value_type;
+  using BoundaryFeatureType = BoundaryFeature<GroupKey, BoundarySubFeatureType>;
+
+  std::vector<BoundaryFeatureType> boundary_features;
+  for (std::size_t i = 0; i < groups.size(); ++i) {
+    for (std::size_t j = i + 1; j < groups.size(); ++j) {
+      const auto next_boundary_features = as_boundary_features(
+          groups[i], groups[j], boundary_function(groups[i], groups[j]));
+      if (next_boundary_features.size() > 0) {
+        boundary_features.insert(boundary_features.end(),
+                                 next_boundary_features.begin(),
+                                 next_boundary_features.end());
+      }
+    }
+  }
+  assert(boundary_features.size() > 0);
+  return boundary_features;
+}
+
+template <typename CovFunc, typename PatchworkFunctions>
+class PatchworkGaussianProcess
+    : public GaussianProcessBase<
+          CovFunc, PatchworkGaussianProcess<CovFunc, PatchworkFunctions>> {
+
+public:
+  using Base = GaussianProcessBase<
+      CovFunc, PatchworkGaussianProcess<CovFunc, PatchworkFunctions>>;
+
+  PatchworkGaussianProcess() : Base(){};
+  PatchworkGaussianProcess(CovFunc &covariance_function)
+      : Base(covariance_function){};
+  PatchworkGaussianProcess(CovFunc &covariance_function,
+                           PatchworkFunctions patchwork_functions)
+      : Base(covariance_function), patchwork_functions_(patchwork_functions){};
+
+  template <typename FitModelType, typename GroupKey>
+  auto
+  from_fit_models(const Grouped<GroupKey, FitModelType> &fit_models) const {
+    using PatchworkFitType = Fit<PatchworkGPFit<FitModelType, GroupKey>>;
+    return FitModel<PatchworkGaussianProcess, PatchworkFitType>(
+        *this, PatchworkFitType(fit_models));
+  };
+
+  template <typename FeatureType, typename FitModelType, typename GroupKey>
+  JointDistribution _predict_impl(
+      const std::vector<FeatureType> &features,
+      const Fit<PatchworkGPFit<FitModelType, GroupKey>> &patchwork_fit,
+      PredictTypeIdentity<JointDistribution> &&) const {
+
+    const auto fit_models = patchwork_fit.fit_models;
+
+    if (fit_models.size() == 1) {
+      return fit_models.values()[0].predict(features).joint();
+    }
+
+    auto get_obs_vector = [](const auto &fit_model) {
+      // TOOD: should these be converted to Measurement<> types?
+      return fit_model.predict(fit_model.get_fit().train_features).mean();
+    };
+    const auto obs_vectors = patchwork_fit.fit_models.apply(get_obs_vector);
+
+    auto boundary_function = [&](const GroupKey &x, const GroupKey &y) {
+      return patchwork_functions_.boundary(x, y);
+    };
+
+    const auto boundary_features =
+        build_boundary_features(boundary_function, fit_models.keys());
```

Comment:
I'll try again ... but there is some trickiness about needing the method to be static if you want it to be treated like a free function. https://stackoverflow.com/questions/12662891/how-can-i-pass-a-member-function-where-a-free-function-is-expected

Reply:
Unfortunately we do need the lambda, using a class method requires the method to be static.  There may be a better way to organize passing these functions around, for example we could pass in the `boundary`, `grouper` and `nearest_group` methods independently (instead of having them bundled in a class) but that introduces more template parameters and I'm not sure what's better.